### ISSUE TITLE
Multiple background AppleScripts don't run in parallel

### ISSUE THREAD

**Issue by insanityfarm** [folivoraAI/BetterTouchTool#851]
**IMPORTANT:** Don't post private requests here (e.g. license issues, etc.). Please send me an email at boastr.net@gmail.com instead.

---

#### Description of bug/feature request/question:

I created an AppleScript that waits for the song currently playing in iTunes to finish, then assigns it a rating. I have five versions of the script (one for each 1-5 star rating) set up as background scripts in BTT and mapped to my function keys F1-F5. The script is written in such a way that I should be able to run it multiple times while a single song plays, and only the last rating requested would be applied when the song ends.

In order to achieve this behavior, BTT would need to allow multiple instances of the script process to run in parallel. I was under the impression the "Run Apple Script in background" action worked this way, but that doesn't seem to be the case.

Instead it queues the script execution so the second one will not begin until the first completes. For my script that means subsequent requests won't even be handled until after the song ends, and will be applied to the next song in the playlist instead. If that's not clear, here's a quick illustration...

###### SCENARIO
Song 1 of my playlist is playing. I press F4, then F5, then F3.

###### EXPECTED RESULT
When Song 1 finishes playing, it is rated three stars.

###### ACTUAL RESULT
When Song 1 finishes playing, it is rated four stars. Then Song 2 plays and is rated five stars. Then Song 3 plays and is rated three stars.

#### Affected input device (e.g. MacBook Trackpad, Magic Mouse/Trackpad, Touch Bar, etc.):

N/A

#### Device information:

Type of Mac: MacBook Pro (Retina, 13-inch, Early 2015)
macOS version: 10.12.4 (16E195)
BetterTouchTool version: 2.094 (620)

#### Additional information (e.g. StackTraces, related issues, screenshots, workarounds, etc.):

It's not clear to me if this is a bug in the way BTT runs AppleScript, or a limitation of AppleScript itself. I've been searching for a way to alter my script so that it spawns a detached process and immediately terminates itself, but I haven't yet had much success. Attaching my AppleScript in case it's helpful: [Rate Current Song 5 Stars.txt](https://github.com/fifafu/BetterTouchTool/files/966443/Rate.Current.Song.5.Stars.txt)

---

**Note:** Before bug reporting, please make sure you have the latest version of BetterTouchTool and that you have already tried to restart your system :-). If you encounter a crash, please attach a crash log from the macOS Console.app from the "User Diagnostic Reports" section.

**Comment by fifafu** [comment]
Unfortunately Apple Scripts are automatically queued by the system when using the cocoa APIs. I'll soon experiment with creating additional processss to run scripts, however I am also not sure whether this will solve the issue. I'll post updates to this here.

**Comment by insanityfarm** [comment]
Thank you! That's unfortunate but I'll monitor the thread for any news of your findings. 😁

**Comment by insanityfarm** [comment]
Update on a workaround I came up with, if you or anyone else are interested...

Running AppleScript asynchronously would've been the holy grail, but all my project really needed was the ability to terminate the instance currently running and then execute it again. I was able to achieve this by invoking the AppleScript from a shell script. First that script uses `pkill` to interrupt any existing processes matching the name of my AppleScript, then it uses `osascript` to start a new one. That command is appended with `&` so bash will run the process in the background, allowing the shell script to exit. I trigger the shell script from BTT.

A side benefit of doing it this way is that I can parameterize the rating to apply to the song in iTunes. Now I don't need to have multiple copies of the same script with different ratings hard-coded into them. Much more elegant all around.

I put it up on [GitHub](https://github.com/insanityfarm/wait-to-rate) in case anyone else might find it useful.
